src/Algebra/Ordered/Structures.agda

```diff
@@ -0,0 +1,308 @@
+
+  isCommutativeMonoid : IsCommutativeMonoid ∙ ε
+  isCommutativeMonoid = record { isMonoid = isMonoid ; comm = comm }
+
+  open IsCommutativeMonoid isCommutativeMonoid public
+    using (isCommutativeSemigroup)
+
+-- Preordered semirings (prosemirings)
+
+record IsProsemiring (+ * : Op₂ A) (0# 1# : A) : Set (a ⊔ ℓ₁ ⊔ ℓ₂) where
+  field
+    +-isCommutativePromonoid : IsCommutativePromonoid + 0#
+    *-cong                   : Congruent₂ *
+    *-mono                   : Monotonic₂ *
+    *-assoc                  : Associative *
+    *-identity               : Identity 1# *
+    distrib                  : * DistributesOver +
+    zero                     : Zero 0# *
+
+  open IsCommutativePromonoid +-isCommutativePromonoid public
+    renaming
+    ( assoc                  to +-assoc
+    ; ∙-cong                 to +-cong
+    ; ∙-congˡ                to +-congˡ
+    ; ∙-congʳ                to +-congʳ
+    ; mono                   to +-mono
+    ; identity               to +-identity
+    ; identityˡ              to +-identityˡ
+    ; identityʳ              to +-identityʳ
+    ; comm                   to +-comm
+    ; isMagma                to +-isMagma
+    ; isSemigroup            to +-isSemigroup
+    ; isMonoid               to +-isMonoid
+    ; isCommutativeSemigroup to +-isCommutativeSemigroup
+    ; isCommutativeMonoid    to +-isCommutativeMonoid
+    )
+
+  *-isPromonoid : IsPromonoid * 1#
+  *-isPromonoid = record
+    { isProsemigroup = record
+      { isPromagma   = record
+        { isPreorder = isPreorder
+        ; ∙-cong     = *-cong
+        ; mono       = *-mono
+        }
+      ; assoc        = *-assoc
+      }
+    ; identity       = *-identity
+    }
+
+  open IsPromonoid *-isPromonoid public
+    using ()
+    renaming
+    ( ∙-congˡ     to *-congˡ
+    ; ∙-congʳ     to *-congʳ
+    ; identityˡ   to *-identityˡ
+    ; identityʳ   to *-identityʳ
+    ; isMagma     to *-isMagma
+    ; isSemigroup to *-isSemigroup
+    ; isMonoid    to *-isMonoid
+    )
+
+  isSemiring : IsSemiring + * 0# 1#
+  isSemiring = record
+    { isSemiringWithoutAnnihilatingZero = record
+      { +-isCommutativeMonoid           = +-isCommutativeMonoid
+      ; *-cong                          = *-cong
+      ; *-assoc                         = *-assoc
+      ; *-identity                      = *-identity
+      ; distrib                         = distrib
+      }
+    ; zero                              = zero
+    }
+
+  open IsSemiring isSemiring public using (distribˡ; distribʳ; zeroˡ; zeroʳ)
+
+------------------------------------------------------------------------
+-- Partially ordered structures
+
+-- Partially ordered magmas (pomagmas)
+
+record IsPomagma (∙ : Op₂ A) : Set (a ⊔ ℓ₁ ⊔ ℓ₂) where
+  field
+    isPartialOrder : IsPartialOrder _≈_ _≤_
+    mono           : Monotonic₂ ∙
+
+  open IsPartialOrder isPartialOrder public
+
+  ∙-cong : Congruent₂ ∙
+  ∙-cong = mono₂⇒cong₂ _≈_ _≈_ Eq.sym reflexive antisym mono
```

Comment:
There's always an argument with these derived properties that they should appear in an auxiliary constructor rather than in the main record. In this case, I'm not really sure; I think we often have a proof of `∙-cong` lying around anyway, but maybe we don't actually gain anything from putting it in because we never really have to look at these proof terms. This could come up if we want to take the opposite ordering and have that be a definitional involution, but maybe because `mono` should stay the same, `∙-cong` does too.

Reply:
Yes, I was thinking along the same lines. @MatthewDaggitt actually suggested a different design where the algebraic structures (rather than the orders) are used as a base, and then there's a better case for keeping `cong` and adding the type of constructors you mention (see our discussion further up in the thread). I'd be curious to hear which design you'd prefer.